### PR TITLE
refactor: add AmuletModel base class, shared ML utilities, and modernize tooling

### LINKED ISSUE #26: Add Abstract Classes for Modules
Most modules in their base class currently either have:
1. General attributes
2. No function definitions

As a result of this, the poisoning base class is essentially empty right now. 

We want to add Abstract Classes (or something similar) for each module with a function definition that must be implemented for that class.

### PR DESCRIPTION
**Pull request by asim29**
## Context

This is PR 1 of a stacked series that splits a large refactor into reviewable layers. No risk-module logic is changed here — this PR lands only the infrastructure that later PRs will import.

**Stack order:**
1. **This PR** — model base class, shared utilities, tooling
2. PR 2 — abstract base classes for all risk modules (API contract)
3. PR 3 — dataset loader rewrites (CelebA, LFW, UTKFace, Census)
4. PRs 4.1–4.7 — one risk vertical per PR (attack + defense + metrics + tests)
5. PR 5 — examples reconciled with new APIs
6. PR 6 — documentation overhaul

## Models (`amulet/models/`)

- **`base.py` (new):** `AmuletModel` — a lightweight `nn.Module` subclass that declares `get_hidden(self, x) -> Tensor`. All existing architectures (VGG, CNN, LinearNet, ResNet) now subclass it. Any PyTorch model can be integrated into Amulet pipelines by wrapping it in a thin `AmuletModel` subclass and implementing `get_hidden` to expose whichever intermediate layer is relevant — no need to reimplement the full model. The architectures in `amulet/models/` are convenience models for experimentation and are not required by the library. Omitting `get_hidden` raises `NotImplementedError` at runtime; several attack modules depend on intermediate activations.

## New shared utilities (`amulet/utils/`)

- **`__meta_classifiers.py` (new):** `PermInvModel` — permutation-invariant meta-classifier used by the White-Box PIM distribution-inference attack (landing in PR 4.1). Placed in `utils/` so any future white-box attack can reuse it without a cross-module import.
- **`__weight_extraction.py` (new):** helpers that extract weight tensors from `nn.Module` instances into the shape `PermInvModel` expects.

## Updated shared utilities (`amulet/utils/`)

`__base.py`, `__pipeline.py`, `__metrics.py`, `__init__.py`:
- Google-style docstring sweep
- Modern type annotations (`Optional[X]` → `X | None`, `typing.List` → `list`)
- `torch.cuda.amp` → `torch.amp` migration (deprecated in PyTorch 2.3)
- `ValueError` raised for unknown datasets instead of `sys.exit()`
- `initialize_model` return type narrowed from `nn.Module` to `AmuletModel`

## Tooling

- **`pyproject.toml`:** ruff config added (`preview = true`, `target-version = "py311"`, rulesets `E W F I UP B C4 SIM RUF`); `pytest-timeout` added; `torch` floor bumped to 2.3.0. `ucimlrepo` is retained here because `amulet/datasets/__tabular_datasets.py` still uses it — it is removed in PR 3 when that file is replaced.
- **`.pre-commit-config.yaml`:** markdownlint and prettier hooks added; ruff hook rev synced with `ruff==` dev dep (a mismatch silently skips rules).
- **`.markdownlint.json`:** per-repo markdownlint config.
- **`.markdownlintignore`:** excludes `docs/` and `examples/extending_amulet/` — these have pre-existing violations that are fixed in PR 6.
- **`AGENTS.md` (new):** minimal agent-guidance file — commands, directory index, and non-obvious invariants only.

## Tests (all new)

- `tests/conftest.py`: shared fixtures including a GPU-parametrised `device` fixture.
- `tests/unit/test_dataset_dataclass.py`: `AmuletDataset` field validation.
- `tests/unit/test_pipeline_{createdir,init,split}.py`: `load_data` / `initialize_model` / split helpers.
- `tests/unit/test_metrics_shared.py`: `compute_accuracy` edge cases.
- `tests/unit/test_meta_classifiers.py`: `PermInvModel` forward pass and output shape.
- `tests/unit/test_weight_extraction.py`: weight-extraction shape contracts.

## Ruff sweep

Adding ruff config exposed 21 pre-existing violations in files not otherwise modified here. These are fixed in this PR to keep CI green. The affected files are rewritten in their respective vertical PRs (datasets in PR 3, risk modules in PRs 4.x, examples in PR 5) — treat these changes as a mechanical style fix accompanying the linter addition.

Closes #26

### CODE REVIEW

**Comment by asim29** [comment]
Includes PR #70

File: .github/workflows/precommit-main.yaml [hunk 1]
```diff
@@ -1,13 +1,12 @@
...
 
-name: precommit-checks
+name: ci
 
 on:
   push:
     branches: [main]
   pull_request:
-    branches: [main, "release/*"]
+    branches: [main, "release/*", "pr/*"]
```
Comment (sebszyller) [inline .github/workflows/precommit-main.yaml:9]: why is this needed? The rule says run CI on branches: main, those mathing release/* and now on branches matching pr/*. If that's what you want, then fine but "pr/*" is a weird naming convention for branches

File: amulet/membership_inference/attacks/lira.py [hunk 2]
```diff
@@ -241,7 +243,7 @@ def __lira_online(
         final_preds = []
         true_labels = []
 
-        for ans, sc in zip(target_in_out_labels, target_scores):
+        for ans, sc in zip(target_in_out_labels, target_scores, strict=False):
```
Comment (sebszyller) [inline amulet/membership_inference/attacks/lira.py]: what does it mean?

File: tests/unit/test_dataset_dataclass.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import numpy as np
+import torch
+from torch.utils.data import TensorDataset
+
+from amulet.datasets.__data import AmuletDataset
+
+
+def _make_datasets() -> tuple[TensorDataset, TensorDataset]:
```
Comment (sebszyller) [inline tests/unit/test_dataset_dataclass.py]: you'd make this into fixtures

File: tests/unit/test_metrics_shared.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader, TensorDataset
+
+from amulet.models import AmuletModel
+from amulet.utils.__base import get_intermediate_features
+from amulet.utils.__metrics import get_accuracy, get_fidelity
+
+
+class MockModel(nn.Module):
```
Comment (sebszyller) [inline tests/unit/test_metrics_shared.py:12]: it'd also be a fixture

File: tests/unit/test_metrics_shared.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    loader = DataLoader(dataset, batch_size=2)
+
+    # Act
+    fid = get_fidelity(m1, m2, loader, cpu_device)
+
+    # Assert
+    assert fid == 0.0
+
+
+class _HiddenMLP(AmuletModel):
```
Comment (sebszyller) [inline tests/unit/test_metrics_shared.py]: maybe also?

**Review by sebszyller — CHANGES_REQUESTED** [review]
It's a kitchen-sink PR so it's possible I missed something but seems mostly fine overall.

File: .github/workflows/precommit-main.yaml (hunk 1 above)
Comment (asim29) [reply .github/workflows/precommit-main.yaml:9]: This was added because all the PRs here are called `pr` which, to be fair, is a name that will likely never be used because I don't think I will ever even _want_ to do a refactor as big as this. Could probably remove it.

File: amulet/membership_inference/attacks/lira.py (hunk 2 above)
Comment (asim29) [reply amulet/membership_inference/attacks/lira.py]: oof this might've been Claude trying to get through https://docs.astral.sh/ruff/rules/zip-without-explicit-strict/ 

good catch, if anything this should be True

File: tests/unit/test_dataset_dataclass.py (hunk 3 above)
Comment (asim29) [reply tests/unit/test_dataset_dataclass.py]: you're right. missed that everywhere, there's likely many places where this is relevant

File: .github/workflows/precommit-main.yaml (hunk 1 above)
Comment (asim29) [reply .github/workflows/precommit-main.yaml:9]: Yeah I'll remove this in the _final_ PR in this godforsaken stack

**Review by sebszyller — APPROVED** [review]
